### PR TITLE
Subscription Management: Rework tracks foundation, add tracks event for subscribe, unsubscribe, site_url/icon/title clicked.

### LINKED ISSUE #78487: Subscription Management: Record tracks event when clicking on site title, site icon and site url.

### PR DESCRIPTION
**Pull request by yansern**
Fixes https://github.com/Automattic/wp-calypso/issues/78487

## Proposed Changes
* Introduce a base tracks event function:
  * recordSubscriptionsTracksEvent( name, props )
* All functions to record tracks event are hooks.
  * The reason for this design is to be able to fetch data returned by hooks.
* All tracks event have the following props (for closer parity with Reader's version):
  * portal = identifies the portal where the event is recorded, e.g. `reader` or `subscriptions`
  * source = (optional) identifies the ui where the event is recorded, e.g. `recommended-site-dismiss-button`, `subscriptions-site-list`
  * subscription_count = user's total subscription count (seems to exist in all reader tracks event, so we're keeping it here too).
* Adds the following tracks event:
  * calypso_subscriptions_site_subscribed
  * calypso_subscriptions_site_unsubscribed
  * calypso_subscriptions_site_resubscribed
  * calypso_subscriptions_site_url_clicked
  * calypso_subscriptions_site_icon_clicked
  * calypso_subscriptions_site_title_clicked
* Reworked previous tracks events:
  * calypso_subscriptions_notifications_toggle_on
  * calypso_subscriptions_notifications_toggle_off
  * calypso_subscriptions_post_emails_toggle_on
  * calypso_subscriptions_post_emails_toggle_off
  * calypso_subscriptions_comment_emails_toggle_on
  * calypso_subscriptions_comment_emails_toggle_off
  * calypso_subscriptions_post_emails_set_frequency
* Some tracks event contains additional reader-related operations (bumpStats, recordGaEvent, etc.)

## Testing Instructions

* Open up `/subscriptions`.
* Open up network panel and filter to `*.gif`.
* Clicking on site title, site icon & site url should log tracks event.
* Changing all site subscription settings should log tracks event.
* Subscribing, unsubscribing and resubscribing should log tracks event.

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/landing/subscriptions/tracks/index.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+export { default as useRecordSiteIconClicked } from './use-record-site-icon-clicked';
+export { default as useRecordSiteTitleClicked } from './use-record-site-title-clicked';
+export { default as useRecordSiteUrlClicked } from './use-record-site-url-clicked';
+export { default as useRecordCommentEmailsToggle } from './use-record-comment-emails-toggle';
+export { default as useRecordNotificationsToggle } from './use-record-notifications-toggle';
+export { default as useRecordPostEmailsSetFrequency } from './use-record-post-emails-set-frequency';
+export { default as useRecordPostEmailsToggle } from './use-record-post-emails-toggle';
+// export { default as useRecordRecommendedSiteSubscribed } from './use-record-recommended-site-subscribed';
+// export { default as useRecordRecommendedSiteUnsubscribed } from './use-record-recommended-site-unsubscribed';
+// export { default as useRecordRecommendedSiteDismissed } from './use-record-recommended-site-dismissed';
```
Comment (yansern) [inline client/landing/subscriptions/tracks/index.ts:14]: Those commented out will be in separate PRs.

File: client/landing/subscriptions/components/site-subscriptions-manager/site-row.tsx [hunk 2]
```diff
@@ -101,15 +112,38 @@ const SiteRow = ( {
...
+	const recordSiteUrlClicked = useRecordSiteUrlClicked();
+	const recordNotificationsToggle = useRecordNotificationsToggle();
+	const recordPostEmailsToggle = useRecordPostEmailsToggle();
+	const recordCommentEmailsToggle = useRecordCommentEmailsToggle();
+	const recordPostEmailsSetFrequency = useRecordPostEmailsSetFrequency();
+	const recordSiteUnsubscribed = useRecordSiteUnsubscribed();
+	const recordSiteResubscribed = useRecordSiteResubscribed();
+
+	// Make object assignment a little easier
+	const SOURCE_SUBSCRIPTIONS_SITE_LIST = 'subscriptions-site-list';
```
Comment (mashikag) [inline client/landing/subscriptions/components/site-subscriptions-manager/site-row.tsx:127]: It is a good practice to define such constants outside the component's render function.

File: client/landing/subscriptions/components/site-subscriptions-manager/site-row.tsx [hunk 3]
```diff
@@ -101,15 +112,38 @@ const SiteRow = ( {
...
+	const blog_id = blog_ID;
+	const feed_id = feed_ID;
```
Comment (mashikag) [inline client/landing/subscriptions/components/site-subscriptions-manager/site-row.tsx:129]: We could follow the same practice we already do for the URL prop, i.e. to rename the prop in the destructuring statement, on line 81.

**Comment by mashikag** [comment]
I tested the PR and all works as expected. I added some minor comments related to code style, but I would not consider them blocking the PR. 👌

**Comment by yansern** [comment]
> I tested the PR and all works as expected. I added some minor comments related to code style, but I would not consider them blocking the PR. 👌

Thanks. I'll address the minor stuff in another PR as I would need to rebase the other PR if I change things in this PR.

**Review by mashikag — APPROVED** [review]

**Comment by yansern** [comment]
> I tested the PR and all works as expected. I added some minor comments related to code style, but I would not consider them blocking the PR. 👌

Addressed here.
https://github.com/Automattic/wp-calypso/pull/78601/commits/1d9662d726b65a13be7465df069020a60f95682d
